Client/game_sa/gamesa_init.h

```diff
@@ -104,3 +104,45 @@ void MemOrFast(U ptr, const T value)
 
 bool GetDebugIdEnabled(uint uiDebugId);
 void LogEvent(uint uiDebugId, const char* szType, const char* szContext, const char* szBody, uint uiAddReportLogId = 0);
+
+struct __THISCALL{};
+struct __CDECL{};
+struct __STDCALL{};
+struct __FASTCALL{};
+struct __VECTORCALL{};
+
+template <typename... Args>
+struct GTAFuncSignature
+{
+    std::tuple<Args...> args;
+    GTAFuncSignature(Args... a) : args(std::forward<Args>(a)...) {}
+};
+
+template <typename... Args>
+GTAFuncSignature<Args...> PrepareSignature(Args... args)
+{
+    return GTAFuncSignature<Args...>(std::forward<Args>(args)...);
+}
+
+template <typename ReturnType, typename Func, typename Tuple, std::size_t... I>
+ReturnType apply_helper(Func&& f, Tuple&& t, std::index_sequence<I...>)
+{
+    return f(std::get<I>(std::forward<Tuple>(t))...);
+}
+
+template <typename ReturnType, typename CallingConvention, typename Func, typename... Args>
+ReturnType CallGTAFunction(Func function, GTAFuncSignature<Args...>& sig)
+{
+    if constexpr (std::is_same_v<CallingConvention, __THISCALL>)
+        return apply_helper<ReturnType>(reinterpret_cast<ReturnType(__thiscall*)(Args...)>(function), sig.args, std::index_sequence_for<Args...>{});
+    else if constexpr (std::is_same_v<CallingConvention, __CDECL>)
+        return apply_helper<ReturnType>(reinterpret_cast<ReturnType(__cdecl*)(Args...)>(function), sig.args, std::index_sequence_for<Args...>{});
+    else if constexpr (std::is_same_v<CallingConvention, __STDCALL>)
+        return apply_helper<ReturnType>(reinterpret_cast<ReturnType(__stdcall*)(Args...)>(function), sig.args, std::index_sequence_for<Args...>{});
+    else if constexpr (std::is_same_v<CallingConvention, __FASTCALL>)
+        return apply_helper<ReturnType>(reinterpret_cast<ReturnType(__fastcall*)(Args...)>(function), sig.args, std::index_sequence_for<Args...>{});
+    else if constexpr (std::is_same_v<CallingConvention, __VECTORCALL>)
+        return apply_helper<ReturnType>(reinterpret_cast<ReturnType(__vectorcall*)(Args...)>(function), sig.args, std::index_sequence_for<Args...>{});
+    else
+        static_assert(sizeof(char) == 0, "Invalid calling convention specified in CallGTAFunction");
```

Comment:
```c++
template <typename ReturnType, typename CallingConvention, typename Func, typename... Args>
ReturnType CallGTAFunction(Func function, Args... params)
{
    if constexpr (std::is_same_v<CallingConvention, __THISCALL>)
        return reinterpret_cast<ReturnType(__thiscall*)(Args...)>(function)(params...);
    else if constexpr (std::is_same_v<CallingConvention, __CDECL>)
        return reinterpret_cast<ReturnType(__cdecl*)(Args...)>(function)(params...);
    else if constexpr (std::is_same_v<CallingConvention, __STDCALL>)
        return reinterpret_cast<ReturnType(__stdcall*)(Args...)>(function)(params...);
    else if constexpr (std::is_same_v<CallingConvention, __FASTCALL>)
        return reinterpret_cast<ReturnType(__fastcall*)(Args...)>(function)(params...);
    else if constexpr (std::is_same_v<CallingConvention, __VECTORCALL>)
        return reinterpret_cast<ReturnType(__vectorcall*)(Args...)>(function)(params...);
    else
        static_assert(sizeof(char) == 0, "Invalid calling convention specified in CallGTAFunction");
}
```
Why do you need `GTAFuncSignature` ?

Reply:
The apply_helper function is equivalent to std::apply to unpack std::tuple. Without this, arguments are passed incorrectly and crashes occur. In the case you mentioned, I would have to use std::make_tuple

Reply:
> Without this, arguments are passed incorrectly and crashes occur.

Can you show an example?

Reply:
In CFxSA.cpp
```cpp
        // Call FxSystem_c::AddParticle
        auto args = PrepareSignature(fxParticleSystem, &vecPosition, &newDirection, 0.0f, &fxPrt, -1.0f, fBrightness, 0.0f, 0);
        CallGTAFunction<int, __THISCALL>(FUNC_FXSystem_c_AddParticle, args);
```

Then try calling this function in your game
```
crun fxCreateParticle("sand", me.position.x, me.position.y, me.position.z, 0, 0, 1, 56, 191, 52, 255, false, 5, 1, 1, true)
```
I haven't analyzed exactly what's wrong, but the arguments are passed in the wrong way, the registers are empty and the game crashes.

Reply:
> The apply_helper function is equivalent to std::apply to unpack std::tuple. Without this, arguments are passed incorrectly and crashes occur. In the case you mentioned, I would have to use std::make_tuple

I removed `PrepareSignature` there. So it's just
```c++
 CallGTAFunction<int, __THISCALL>(FUNC_FXSystem_c_AddParticle, fxParticleSystem, &vecPosition, &newDirection, 0.0f, &fxPrt, -1.0f, fBrightness, 0.0f, 0);
 ```

Reply:
Lol, due to all the confusion and the multitude of proposals, I forgot about it. Yeah, that's great